### DISCUSSION TITLE
[shipped] course data is now automatically synced!

### DISCUSSION THREAD

**Discussion by qaisjp** [qaisjp/inf1.hgs.club#102] [Show and tell]
TL;DR: As of https://github.com/compsoc-edinburgh/betterinformatics/commit/e1ba9bbe7c2be26f991ff36e2c81dd782d825162 we now automatically sync course data from [course.inf.ed.ac.uk](https://course.inf.ed.ac.uk) to [`master/_data/courses.yml`](https://github.com/compsoc-edinburgh/betterinformatics/blob/master/_data/courses.yml)!

**The what file?**

The `courses.yaml` file is a dump of Informatics' course listing. Better Informatics uses it to organise courses into semesters, suggest relevant courses (third years can take fourth year courses too, as long as they are Level 10), and probably other stuff too.

**Explanation**

Previously, whenever I noticed the course data was a bit stale, I'd run the courses.yaml generator that lives in [bi-app](https://github.com/compsoc-edinburgh/bi-app). Then I'd copy the YAML file it generated and paste it over the YAML file in this repo. All of this is manual work.

Now, every weekday at 7pm, that script and [this GitHub workflow](https://github.com/compsoc-edinburgh/betterinformatics/blob/master/.github/workflows/courses.yml) will automatically commit and push any `courses.yaml` updates to `master`, completely on its own!

It won't create extra commits, either. If there are no real changes, it will simply do nothing!

**What if it changes mid-week and I want it to update NOW!**

That's cool. You can also run it manually. Just follow these instructions:

1. Click the _Actions_ tab
2. Select the _Update courses.yaml data_ workflow
3. Click _Run workflow_
4. Leave _Branch_ as _master_, and click _Run workflow_

![image](https://user-images.githubusercontent.com/923242/92824136-f8699f80-f3c5-11ea-8a6c-846b8e9c31e9.png)

**The nitty gritty**

To make this possible, there were [a few changes](https://github.com/compsoc-edinburgh/bi-app/compare/e41e94d9084413407476b746810d4d9316daf374...afcb7a3406f9926f5999ed7b6202d5ce5588329e) I had to make to the YAML generator:

- Remove hardcoded academic years.
- Add some diffing code that would print out a special directive that GitHub workflows understand https://github.com/compsoc-edinburgh/bi-app/blob/afcb7a3406f9926f5999ed7b6202d5ce5588329e/courses.yaml/app.py#L171-L177, so that we don't create commits when nothing has actually changed.

You can view the first ever run here: https://github.com/compsoc-edinburgh/betterinformatics/runs/1099186437?check_suite_focus=true

-----

I'm happy to answer any and all questions!

(I'm leaving this announcement open for a little while, and I will close it later.)
